File: src/CodeCracker/SealedAttributeAnalyzer.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Diagnostics;
using System;
using System.Collections.Immutable;

namespace CodeCracker
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class SealedAttributeAnalyzer : DiagnosticAnalyzer
    {
        public const string DiagnosticId = "CC0023";
        internal const string Title = "Unsealed Attribute";
        internal const string MessageFormat = "Mark '{0}' as sealed.";
        internal const string Category = "Performance";
        const string Description = "Framework methods that retrieve attributes by default search the whole "
            + "inheritence hierarchy of the attribute class. "
            + "Marking the type as sealed eliminate this search and can improve performance";

        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId,
            Title,
            MessageFormat,
            Category,
            DiagnosticSeverity.Warning,
            true,
            description: Description,
            helpLink: HelpLink.ForDiagnostic(DiagnosticId));

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterSymbolAction(Analyze, SymbolKind.NamedType);
        }

        private void Analyze(SymbolAnalysisContext context)
        {
            var type = (INamedTypeSymbol)context.Symbol;

            if (type.TypeKind != TypeKind.Class) return;

            if (!IsAttribute(type)) return;

            if (type.IsAbstract || type.IsSealed) return;

            context.ReportDiagnostic(Diagnostic.Create(Rule, type.Locations[0], type.Name));
        }

        public static bool IsAttribute(ITypeSymbol symbol)
        {
            var @base = symbol.BaseType;
            var attributeName = typeof(Attribute).Name;

            while (@base != null)
            {
                if (@base.Name == attributeName)
                    return true;

                @base = @base.BaseType;
            }

            return false;
        }
    }
}

File: src/VisualBasic/CodeCracker/Performance/SealedAttributeAnalyzer.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports System.Collections.Immutable
Imports CodeCracker.Extensions

Namespace Performance
    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
    Public Class SealedAttributeAnalyzer
        Inherits DiagnosticAnalyzer

        Public Shared ReadOnly Id As String = DiagnosticId.SealedAttribute.ToDiagnosticId()
        Public Const Title As String = "Unsealed Attribute"
        Public Const MessageFormat As String = "Mark '{0}' as NotInheritable."
        Public Const Category As String = SupportedCategories.Performance
        Public Const Description As String = "Framework methods that retrieve attributes by default search the entire inheritence hierarchy of the attribute class. Marking the type as NotInheritable eliminates this search and can improve performance."
        Protected Shared Rule As DiagnosticDescriptor = New DiagnosticDescriptor(
            Id,
            Title,
            MessageFormat,
            Category,
            DiagnosticSeverity.Warning,
            isEnabledByDefault:=True,
            description:=Description,
            helpLink:=HelpLink.ForDiagnostic(Id))

        Public Overrides ReadOnly Property SupportedDiagnostics() As ImmutableArray(Of DiagnosticDescriptor) = ImmutableArray.Create(Rule)

        Public Overrides Sub Initialize(context As AnalysisContext)
            context.RegisterSymbolAction(AddressOf Analyze, SymbolKind.NamedType)
        End Sub

        Private Sub Analyze(context As SymbolAnalysisContext)
            Dim type = DirectCast(context.Symbol, INamedTypeSymbol)
            If type.TypeKind <> TypeKind.Class Then Exit Sub

            If Not IsAttribute(type) Then Exit Sub
            If (type.IsAbstract OrElse type.IsSealed) Then Exit Sub
            context.ReportDiagnostic(Diagnostic.Create(Rule, type.Locations(0), type.Name))
        End Sub

        Public Shared Function IsAttribute(symbol As ITypeSymbol) As Boolean
            Dim base = symbol.BaseType
            Dim attributeName = GetType(Attribute).Name
            While base IsNot Nothing
                If base.Name = attributeName Then Return True
                base = base.BaseType
            End While
            Return False
        End Function

    End Class
End Namespace